kvbc/src/categorization/kv_blockchain.cpp

```diff
@@ -240,4 +247,78 @@ ImmutableUpdatesInfo KeyValueBlockchain::handleCategoryUpdates(
+  const auto last_reachable_block = getLastReachableBlockId();
+  if (block_id <= last_reachable_block) {
+    const auto msg = "Cannot add an existing block ID " + std::to_string(block_id);
+    throw std::invalid_argument{msg};
+  }
+
+  auto write_batch = native_client_->getBatch();
+  state_transfer_block_chain_.addBlock(block_id, block, write_batch);
+  native_client_->write(std::move(write_batch));
+  // Update the cached latest ST temporary block ID if we have received and persisted such a block.
+  state_transfer_block_chain_.updateLastId(block_id);
+
+  try {
+    linkSTChainFrom(last_reachable_block + 1);
+  } catch (const std::exception& e) {
+    // LOG_FATAL(logger_, "Aborting due to failure to link chains after block has been added, reason: "s + e.what());
+    std::terminate();
+  } catch (...) {
+    // LOG_FATAL(logger_, "Aborting due to failure to link chains after block has been added");
+    std::terminate();
+  }
+}
+
+RawBlock KeyValueBlockchain::getRawBlock(const BlockId& block_id) const {
+  const auto last_reachable_block = getLastReachableBlockId();
+
+  // Try to take it from the ST chain
+  if (block_id > last_reachable_block) {
+    auto raw_block = state_transfer_block_chain_.getRawBlock(block_id);
+    if (!raw_block) {
+      // E.L throw or optional?
+      throw std::runtime_error{"Failed to get block node ID = " + std::to_string(block_id)};
+    }
+    return raw_block.value();
+  }
+
+  // Try from the blockchain itself
+  auto raw_block = block_chain_.getRawBlock(block_id);
+  if (!raw_block) {
+    throw std::runtime_error{"Failed to get block node ID = " + std::to_string(block_id)};
+  }
+  return raw_block.value();
+}
+
+// tries to remove blocks form the state transfer chain to the blockchain
+void KeyValueBlockchain::linkSTChainFrom(BlockId block_id) {
+  const auto last_block_id = state_transfer_block_chain_.getLastBlockId();
+  if (!last_block_id) return;
+
+  for (auto i = block_id; i <= last_block_id.value(); ++i) {
+    auto raw_block = state_transfer_block_chain_.getRawBlock(i);
+    if (!raw_block) {
+      return;
+    }
+    writeSTLinkTransaction(block_id, *raw_block);
+  }
+  // Linking has fully completed and we should not have any more ST temporary blocks left. Therefore, make sure we don't
+  // have any value for the latest ST temporary block ID cache.
+  state_transfer_block_chain_.resetChain();
+}
+
+// Atmoic delete from state transfer and add to blockchain
+void KeyValueBlockchain::writeSTLinkTransaction(const BlockId block_id, RawBlock& block) {
+  auto write_batch = native_client_->getBatch();
+  state_transfer_block_chain_.deleteBlock(block_id, write_batch);
+  auto new_block_id = addBlock(std::move(block.data.category_updates), write_batch);
+  native_client_->write(std::move(write_batch));
+
+  block_chain_.setAddedBlockId(new_block_id);
+  state_transfer_block_chain_.updateLastIdAfterDeletion(block_id);
```

Comment:
I am not sure we need to call `updateLastIdAfterDeletion()` here as we don't change the last ID of the ST chain. Doing so works, but it does a load that in turn seeks and, if I am not missing something, is unnecessary. Please correct me if I am wrong.

Reply:
You're right, in this context is redundant, I'll remove it

(Though it won't actually do anything due to the condition at the start 
```if (!last_block_id_.has_value() || last_block_id_.value() != id) {
        return;
   }```)